### PR TITLE
DelegatedRole::matchesPath() returns true for non-matching targets if path hash prefixes are used

### PR DESCRIPTION
**Pull request by phenaproxima**
Major bug!

If a delegated role doesn't match a target when path hash prefixes are used, we...return `true` from `matchesPath()`. That's completely and totally wrong, and it's manifested when we try to get TUF metadata from drupal.org, which has a bunch of terminating hash bin delegations.

What we need to do here is ensure that, if path hash prefixes are used, we return false if there's no match. That would be correct, and in accordance with the spec.

### CODE REVIEW

File: src/DelegatedRole.php [hunk 1]
```diff
@@ -122,7 +122,8 @@ public function matchesPath(string $target): bool
             }
             return false;
```
Comment (tedbow) [inline src/DelegatedRole.php:123]: Why don't we have this same `return false` in the block for `(isset($this->pathHashPrefixes))`

Can a delegated roll have both `pathHashPrefixes` and `paths`?

Comment (phenaproxima) [reply src/DelegatedRole.php:123]: No, it cannot, which is why we now throw `LogicException` at the end of the method. This is also enforced by validation (already existing in HEAD).

File: src/DelegatedRole.php [hunk 2]
```diff
@@ -122,7 +122,8 @@ public function matchesPath(string $target): bool
...
         }
-        // If no paths are set then any target is a match.
-        return true;
+
+        // It should never be possible to get here.
+        throw new \LogicException('No paths or path hash prefixes are defined for this role.');
```
[the remark this answers is not in the document]
Comment (phenaproxima) [reply src/DelegatedRole.php:127]: Good catch; fixed.

File: generate_fixtures.py [hunk 3]
```diff
@@ -39,6 +38,7 @@ def generate_fixtures():
     PublishedTwice.build(rotate_keys='timestamp')
     PublishedTwice.build(rotate_keys='snapshot')
     TargetsLengthNoSnapshotLength.build()
+    HashedBins.build()
```
[the remark this answers is not in the document]
Comment (phenaproxima) [reply generate_fixtures.py:41]: Just debugging. I'll revert it.

**Review by tedbow — APPROVED** [review]
Looks good!

**Pull request merged**